pspy/so_cov.py

```diff
@@ -199,7 +212,93 @@ def cov_coupling_spin0and2_simple(win, lmax, niter=3, save_file=None, planck=Fal
         np.save("%s.npy"%save_file, coupling)
 
     return coupling_dict
+    
+    
+def fast_cov_coupling_spin0and2(sq_win_alms_dir,
+                                id_element,
+                                lmax,
+                                l_exact=None,
+                                l_band=None,
+                                l_toep=None):
+                                
+    """Compute the coupling kernels corresponding to the T and E covariance matrix
+    This routine assume that you have already precomputed the alms of the square of the windows
+    and that the window in temperature and polarisation are the same.
+    Both conditions lead to a very important speed up, that explains the name of the routine
 
+    Parameters
+    ----------
+
+    sq_win_alms_dir: string
+        the folder with precomputed alms corresponding to the sq of the window function
+        this is what enters the analytic computation
+    id_element : list
+        a list of the form [a,b,c,d] where a = dr6_pa4_090, etc, this identify which pair of power spectrum we want the covariance of
+    lmax: integer
+      the maximum multipole to consider
+    l_toep: int
+      parameter for the toeplitz approximation (see arXiv:2010.14344)
+    l_band: int
+      parameter for the toeplitz approximation (see arXiv:2010.14344)
+    l_exact: int
+      parameter for the toeplitz approximation (see arXiv:2010.14344)
+
+    """
+
+    na, nb, nc, nd = id_element
+
+    if l_toep is None: l_toep = lmax
+    if l_band is None: l_band = lmax
+    if l_exact is None: l_exact = lmax
+    
+    def try_alm_order(dir, a, b):
+        try:
+            alm = np.load("%s/alms_%sx%s.npy"  % (dir, a, b))
+        except:
+            alm = np.load("%s/alms_%sx%s.npy"  % (dir, b, a))
+        return alm
+    
+    alm_TaTc = try_alm_order(sq_win_alms_dir, na, nc)
+    alm_TbTd = try_alm_order(sq_win_alms_dir, nb, nd)
+    alm_TaTd = try_alm_order(sq_win_alms_dir, na, nd)
+    alm_TbTc = try_alm_order(sq_win_alms_dir, nb, nc)
+
+    wcl = {}
+    wcl["TaTcTbTd"] = hp.alm2cl(alm_TaTc, alm_TbTd)
+    wcl["TaTdTbTc"] = hp.alm2cl(alm_TaTd, alm_TbTc)
+
+    l = np.arange(len(wcl["TaTcTbTd"]))
```

Comment:
Maybe we could check if the maximum multipole value from `alm2cl` stays above `lmax`, otherwise we manually shrink `lmax` to `np.max(l)` or `l[-1]`